File: 프로그래머스/lv3/49189. 가장 먼 노드/가장 먼 노드.py
from collections import deque

        
def solution(n, edge):
    dist = [-1 for _ in range(n + 1)]
    graph = [[] for _ in range(n + 1)]
    
    for e in edge:
        graph[e[0]].append(e[1])
        graph[e[1]].append(e[0])
    
    q = deque()
    q.append(1)
    
    dist[1] = 1
    
    while q:
        now = q.popleft()
        for next in graph[now]:
            if dist[next] == -1:
                dist[next] = dist[now] + 1
                q.append(next)
    mDist = max(dist)
    answer = 0
    for i in range(1, n + 1):
        if mDist == dist[i]:
            answer += 1

    return answer if answer > 1 else 1

File: 프로그래머스/lv3/49189. 가장 먼 노드/가장 먼 노드.java
import java.util.*;
import java.awt.Point;
class Solution {
    
    public int solution(int n, int[][] edge) {
        boolean[][] graph = new boolean[n + 1][n + 1];
        int[] dist = new int[n + 1];
        
        for(int i = 0; i < n; ++i){
            Arrays.fill(graph[i], false);
        }
        Arrays.fill(dist, -1);
        for(int[] e: edge){
            graph[e[0]][e[1]] = true;
            graph[e[1]][e[0]] = true;
        }
        Queue<Integer> q = new LinkedList<>();
        dist[1] = 0;
        q.add(1);
        int maxDist = 0;
        while(!q.isEmpty()){
            int now = q.poll();
            for(int i = 1; i < n + 1; ++i){
                if(dist[i] == -1 && graph[now][i]){
                    dist[i] = dist[now] + 1;
                    maxDist = Math.max(maxDist, dist[i]);
                    q.add(i);
                }
            }
        }
        int answer = 0;
        for(int i = 1; i < n + 1; ++i)
            if(dist[i] == maxDist)
                answer++;

        return answer;
    }
}